src/borg/locking.py

```diff
@@ -124,30 +125,50 @@ def acquire(self, timeout=None, sleep=None):
             timeout = self.timeout
         if sleep is None:
             sleep = self.sleep
-        timer = TimeoutTimer(timeout, sleep).start()
-        while True:
-            try:
-                os.mkdir(self.path)
-            except FileExistsError:  # already locked
-                if self.by_me():
+        parent_path, base_name = os.path.split(self.path)
+        unique_base_name = os.path.basename(self.unique_name)
+        temp_path = None
+        try:
+            temp_path = tempfile.mkdtemp(".tmp", base_name + '.', parent_path)
+            temp_unique_name = os.path.join(temp_path, unique_base_name)
+            with open(temp_unique_name, "wb"):
+                pass
+        except OSError as err:
+            raise LockFailed(self.path, str(err)) from None
+        else:
+            timer = TimeoutTimer(timeout, sleep).start()
+            while True:
+                try:
+                    os.rename(temp_path, self.path)
+                except OSError:  # already locked
+                    if self.by_me():
+                        return self
+                    self.kill_stale_lock()
+                    if timer.timed_out_or_sleep():
+                        raise LockTimeout(self.path)
+                else:
+                    temp_path = None
                     return self
-                self.kill_stale_lock()
-                if timer.timed_out_or_sleep():
-                    raise LockTimeout(self.path)
-            except OSError as err:
-                raise LockFailed(self.path, str(err)) from None
-            else:
-                with open(self.unique_name, "wb"):
-                    pass
-                return self
+        finally:
+            if temp_path is not None:  # Renaming failed for any reason, so temp_dir still exists and should be cleaned up anyway.
+                os.unlink(temp_unique_name)
+                os.rmdir(temp_path)
 
     def release(self):
         if not self.is_locked():
             raise NotLocked(self.path)
         if not self.by_me():
             raise NotMyLock(self.path)
         os.unlink(self.unique_name)
-        os.rmdir(self.path)
+        try:
+            os.rmdir(self.path)
+        except OSError as err:
+            if err.errno == errno.ENOTEMPTY or err.errno == errno.ENOENT:
+                # Directory is not empty or doesn't exist any more = we lost the race to somebody else--which is ok.
+                pass
+            else:
+                # EACCES or EIO or ... = we cannot operate anyway, so re-throw
+                raise err
```

Comment:
can we invert the condition to save the `pass`?

like:
```
            if err.errno not in (errno.ENOTEMPTY, errno.ENOENT, ):
                # EACCES or EIO or ... = we cannot operate anyway, so re-throw
                raise err
            # else: directory is not empty or doesn't exist any more = we lost the race to somebody else (which is ok).
```

Reply:
done

Reply:
Yeah. You see, I'm a bloody Python beginner, I wasn't thinking in Python...